File: Search/BinarySearch/scala/BinarySearch.scala
/**
  Recursive implementation of Binary Search
  */

object BinarySearch {

  /**
   * Returns the index of the key in the given array if found,
   * otherwise returns -1.
   */
  def search(arr: Array[Int], l: Int, r: Int, key: Int): Int = {
    if (l <= r) {
      val mid = l + (r - l) / 2
      if (key == arr(mid))
        mid
      else if (key < arr(mid))
        search(arr, l, mid - 1, key)
      else
        search(arr, mid + 1, r, key)
    } else -1
  }

  def main(args: Array[String]): Unit = {
    val sortedArray: Array[Int] = Array(10, 20, 30, 40, 50, 60, 70, 80, 90)
    val (l, r): (Int, Int) = (0, sortedArray.length - 1)
    val key: Int = 50
    val result: Int = search(sortedArray, 0, sortedArray.length - 1, key)

    if (result == -1)
      println(s"$key is not found in the array!")
    else
      println(s"$key is found at index $result.")
  }
}

File: Search/BinarySearch/java/BinarySearch.java
/**
 * @author: shakeelsamsu
 * https://github.com/shakeelsamsu
 * Implementation of Binary Search in Java
 * Returns the location of element x in a given array (if present), or otherwise returns -1
 */
class BinarySearch {
    int binarysearch(int[] nums, int x) {
        int i = 0;
        int f = nums.length - 1;
        while(i <= f) {
            
            int mid = i + (f - i) / 2; //find the middle

            //check the middle
            if(nums[mid] == x) return mid;

            //if x is greater than the mid, ignore the left
            if(nums[mid] < x) i = mid + 1;
            
            //otherwise ignore the right
            else f = mid - 1;
        }

        //if we reach here, the element must not be present
        return -1;
    }

    public static void main(String[] args) {
        BinarySearch bs = new BinarySearch();
        int[] arr = {10, 2, 7, 3, 9, 15, 6};
        int n = arr.length;
        int x = 15;
        int result = bs.binarysearch(arr, x);
        if(result == - 1) System.out.println(x + " was not found");
        else System.out.println(x + " was found at index " + result);
    }
}
